packages/core/content-releases/server/src/services/release.ts

```diff
@@ -500,200 +574,176 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => {
+            if (actions.length === 0) {
+              throw new errors.ValidationError('No entries to publish');
+            }
 
-            try {
-              if (action === 'publish') {
-                await entityManagerService.publish(entry, uid);
-              } else {
-                await entityManagerService.unpublish(entry, uid);
-              }
-            } catch (error) {
-              if (
-                error instanceof errors.ApplicationError &&
-                (error.message === 'already.published' || error.message === 'already.draft')
-              ) {
-                // We don't want throw an error if the entry is already published or draft
+            /**
+             * We separate publish and unpublish actions, grouping them by contentType and extracting only their IDs. Then we can fetch more data for each entry
+             * We need to separate collectionTypes from singleTypes because findMany work as findOne for singleTypes and publishMany can't be used for singleTypes
+             */
+            const collectionTypeActions: {
+              [key: UID.ContentType]: {
+                entriesToPublishIds: ReleaseAction['entry']['id'][];
+                entriesToUnpublishIds: ReleaseAction['entry']['id'][];
+              };
+            } = {};
+            const singleTypeActions: {
+              uid: UID.ContentType;
+              id: ReleaseAction['entry']['id'];
+              action: ReleaseAction['type'];
+            }[] = [];
+            for (const action of actions) {
+              const contentTypeUid = action.contentType;
+
+              if (strapi.contentTypes[contentTypeUid].kind === 'collectionType') {
+                if (!collectionTypeActions[contentTypeUid]) {
+                  collectionTypeActions[contentTypeUid] = {
+                    entriesToPublishIds: [],
+                    entriesToUnpublishIds: [],
+                  };
+                }
+
+                if (action.type === 'publish') {
+                  collectionTypeActions[contentTypeUid].entriesToPublishIds.push(action.entry.id);
+                } else {
+                  collectionTypeActions[contentTypeUid].entriesToUnpublishIds.push(action.entry.id);
+                }
               } else {
-                throw error;
+                singleTypeActions.push({
+                  uid: contentTypeUid,
+                  action: action.type,
+                  id: action.entry.id,
+                });
               }
             }
-          }
 
-          // Then, we can continue with publishing the collectionTypes
-          for (const contentTypeUid of Object.keys(collectionTypeActions)) {
-            // @ts-expect-error - populateBuilderService should be a function but is returning service
-            const populate = await populateBuilderService(contentTypeUid)
-              .populateDeep(Infinity)
-              .build();
+            await strapi.db.transaction(async () => {
```

Comment:
Something similar to this was my first idea (you can see it in the first commit of this PR) and it works partially.... 

To understand this, consider the following:
```
const trx = await strapi.db.transaction();

const lockedRelease = strapi.db.queryBuilder()......
  .transacting(trx)
  
publishSingleType()
```
1. I created the transaction.
2. I used it to lock the release's row in the database so all other processes keep waiting for this to be resolved.

Now there are two following scenarios depending on what DB you are using:

**For Postgres/Mysql** (here the locks work as expected, meaning I'm locking ONLY the access to the release's row):
3. When executing the `publishSingleType`, there is no problem because this function changes different tables so everything works as expected.

**For SQLite** (In SQLite there are no row locks, which makes sense because the complete database is locked for any request. So, when you start a transaction in SQLite, you basically lock the entire database):
3. When executing the `publishSingleType`, it tries to change stuff in different tables, but all the other queries don't have the transaction, so when they try to access the database, it's locked.

The nested transaction solved this problem because all calls to strapi.db or entityService inside a transaction are using the transaction's context.